### PR TITLE
Import queue API revamp

### PR DESCRIPTION
**Pull request by tomaka**
cc #2536 #2416 
Can be reviewed commit by commit.

This PR combines a few changes:

- Since the import queue is owned only by a single place, I made all of its trait methods take `&mut self` instead of `&self`. **If there is ever a need to share an import queue between multiple places, please implement the `ImportQueue` trait on `Arc<BasicImportQueue>`** instead of reverting back to `&self`. I also made the methods of `Link` take `&mut`.

- I removed the `on_start` method of the `ImportQueue`. The queue is considered started as soon as we create it. But this raises the question: how do we pass the link to the network? That's where the next change comes into play.

- There's a new `ImportQueue::poll_actions` that gets passed a `&mut dyn Link` and flushes actions that have been enqueued. It behaves like `Future::poll` in the sense that it registers the current task and notifies it. In other words, from the external point of view, it is `poll_actions` that advances the import queue state machine, and the fact that there's a background thread is an implementation detail.

- I removed the `BasicSyncQueue` as it was unused anyway. I know it's me who asked for me, but in the end it would be a lot of efforts to update it, while it's totally useless.

- You might notice that the tests are gone. These tests were testing the channel between the front and background threads, and didn't really make sense to me. We know that channel works.

The consequences of these changes are:

-  In the future we can easily make the `ImportQueue` use a tokio task instead of a background thread. I didn't dare make the change as I think that this kind of change would need to be battle-tested before being merged and deserve their own PR. It also unlocks other possibilities, such as spawning a new task for each block to import.

- Justifications used to be imported in a thread that's separate from the thread that imports blocks and finality proofs. This is no longer the case. My guess is that this was done more because it was the easiest way to implement justifications, and not for any good specific reason.

- The queue of network actions is moved from the network thread to the import queue. In some rare situations, things may happen in a slightly different order than before (e.g. we might wake up the sync processing code a bit later). Normally there's no reason why things would break, and if things do break then it is clearly that there was already a bug somewhere beforehand.

The next changes would be lots of cleanups in the network crate, and replacing all the shared verifiers with non-shared verifiers, for more Arcs and Mutexes removals.

### CODE REVIEW

**Comment by tomaka** [comment]
> The worker then also calls various start methods, passing along the Link implementing sender.

That's clearly not great, and would be removed by a PR on top of this one that would apply the same design to the three `*Import` traits.

> This seems complicated, and having a trait named ImportQueue managing all this seems out of place.

The `ImportQueue` trait is only concerned with `Link`.
The fact that the `BasicImportQueue` uses the `BufferedLinkSender`/`BufferedLinkReceiver` is purely an implementation detail, and is none of the concern of the user of this trait.

If this is too complicated, an alternative could be to make `poll_actions` return a `BlockImportWorkerMsg` (that would be renamed to something else). I don't really have an opinion about this, and went for the option that doesn't break too much.

> Re-introduce the start method on ImportQueue, passing it along the BufferedLinkSender implementing Link.

Just like everything that touches #2536, my objective is to hide the fact that we have threads and channels as implementation details.
In other words, I precisely want to **not** expose the `BufferedLinkSender` (or any channel or anything that involves threading whatsoever) as part of the API of `ImportQueue`.

File: core/consensus/common/src/import_queue.rs [hunk 1]
```diff
@@ -439,6 +454,139 @@ pub trait Link<B: BlockT>: Send {
...
+		who: Origin,
+		request_block: (B::Hash, NumberFor<B>),
+		finalization_result: Result<(B::Hash, NumberFor<B>), ()>,
+	) {
+		let msg = BlockImportWorkerMsg::FinalityProofImported(who, request_block, finalization_result);
+		let _ = self.tx.unbounded_send(msg);
+	}
+
+	fn request_finality_proof(&mut self, hash: &B::Hash, number: NumberFor<B>) {
+		let _ = self.tx.unbounded_send(BlockImportWorkerMsg::RequestFinalityProof(hash.clone(), number));
```
Comment (DemiMarie-parity) [inline core/consensus/common/src/import_queue.rs]: Can we use `drop` here?

Comment (tomaka) [reply core/consensus/common/src/import_queue.rs]: `let _ = ` is the Rust idiomatic thing.

**Comment by tomaka** [comment]
> Is it really just an implementation detail? If you call ImportQueue.poll_actions outside of the context of a task, you risk a panic: https://docs.rs/futures/0.1.25/futures/future/trait.Future.html#panics

That's mentioned in the documentation of `poll_actions`. It is made obvious from the word `poll` which should ring the "async context" bell of the user.

> Instead I suggest having the user of the trait provide a Link, via a start or set_link method, and then have the end-user be responsible for polling the actions on the other side of the Link.

I need some sort of function that requires to be called in an asynchronous context, because `tokio::spawn` requires that. Otherwise we're forced to use either threads as opposed to tasks, or get passed a `TaskExecutor` which would our API to tokio.

Additionally, having `poll_actions` is the ultimate flexibility: you can do anything you want with that.

You can for example process justifications in the main task and blocks in the background task, or use multiple background tasks, without having to put an Arc/Mutex around the link.

You can for example do all the processing in the main thread but in small increments (eg. submit you submit 10 blocks for import, but `poll_actions` processes one block, then goes to sleep 1 second, then processes another one, sleeps, etc. all in the foreground). This is something we might do for the WASM-browser situation.

Very importantly, it also makes it possible to have the import queue in sync with the background worker. If we ever need a `is_block_being_processed` method on `ImportQueue`, we now can do that without running into weird problems! The `BasicImportQueue` would just need a `HashSet<B::Hash>` field or something similar. Try doing that by passing a `Link` to the background and pull out your hair.

Your only argument seems to be "it's complicated", which to me is pretty much not true. From the user's point of view, the import queue is simple a state machine. I could even argue that it's less complicated to use than right now, as right now you have to handle the fact that the method on the `Link` that you pass can be called at any moment.

**Comment by arkpar** [comment]
LGTM in general

**Comment by gterzian** [comment]
> We need some sort of function that requires to be called in an asynchronous context, because tokio::spawn requires that.

`BufferedLinkReceiver.poll`

> Additionally, having poll_actions is the ultimate flexibility for the implementer: you can do anything you want with that.

> You can for example process justifications in the main task and blocks in the background task, or use multiple background tasks, without having to put an Arc/Mutex around the link.

> You can for example do all the processing in the main thread but in small increments (eg. submit you submit 10 blocks for import, but poll_actions processes one block, then goes to sleep 1 second, then processes another one, sleeps, etc. all in the foreground). This is something we might do for the WASM-browser situation.

Ok so now I don't understand what you're trying to do. `poll_actions` seems to have nothing to do with the processing of blocks or justifications, which is done in the actual import queue worker. That worker can keep working at whatever pace it wants, since results will be sent on the `BufferedLinkSender`, which is an unbounded channel. 

`poll_actions` is only a way to receive messages on the `BufferedLinkReceiver`, and then handle those messages via method calls on `Link`. But since the `Link` passed to `poll_actions` is just the `NetworkLink` that is in the same task as the one calling `poll_actions`, you might as well remove this entire back and forth, poll the `BufferedLinkReceiver`, and call methods on protocol or network service. 

What I find too complicated is that you already have one `Link`, which is the `BufferedLinkSender`, corresponding to the `BufferedLinkReceiver` on the other side, and so I see no need to introduce a second `Link`, the `NetworkLink`, and call `import_queue.poll_actions(link)`. It seems like a complicated way to do `BufferedLinkReceiver.poll`.

I agree with your goals of flexibility, and I don't understand the point of `poll_actions` in getting there.

**Comment by folsen** [comment]
Lots of unnecessary discussion here IMO that just confuses the matter and goes in circles.

@tomaka Having the goal stated more clearly in opening the PR would help, e.g. "This is the problem, here is my solution." or generally more clearly establishing goals.

@gterzian You're more or less just saying "this seems complicated" without offering an alternative (other than just, dropping the PR entirely which seems to be what you want to do). A review should be based around "Is the problem valid?", "Does this solve the problem correctly (without logic errors or bugs)?". A PR shouldn't be held up by opinions on style. If you feel there is a much better way to solve the problem you should provide an alternative. After reading all of your comments the only things I can conclude is that you disagree with the PR because it's not how you would do it, or disagree that there is a problem completely. If you disagree that there is a problem, then ask for clarification on the problem, if you disagree with the solution, provide a better one. If you just feel like it doesn't obey your own style of coding, then please provide that input, but let others express their opinions as well. We've had 2 other reviewers that seem to think the style is fine.

**Comment by gterzian** [comment]
> If you feel there is a much better way to solve the problem you should provide an alternative. After reading all of your comments the only things I can conclude is that you disagree with the PR because it's not how you would do it, or disagree that there is a problem completely.

I suggest you read my initial response again, it contains a detailed suggestion, which doesn't involve dropping the entire PR. 

I'll re-iterate here:

- Move the BufferedLinkSender/Receiver code to network.
- Re-introduce the start method on ImportQueue, passing it along the BufferedLinkSender implementing Link.
- call buffered_link() inside NetworkWorker::new, and also call call ImportQueue.start, passing along the BufferedLinkSender.
- Instead of calling poll_actions, simply poll the BufferedLinkReceiver and call methods on Protocol directly to handle messages, without the need for the NetworkLink intermediary.

Other than that I think the PR is fine, I simply dislike what seems like unnecessary complications around the proposed `poll_actions` method. 

It doesn't seem like a question of style, and it goes to the hearth of the PR, whose goal is improving the API of the `ImportQueue`. 

If everyone else feel the use of `poll_actions` is absolutely clear, then so be it. We've discussed it so much by now I'm almost getting used to it...

**Comment by tomaka** [comment]
> Ok so now I don't understand what you're trying to do. poll_actions seems to have nothing to do with the processing of blocks or justifications, which is done in the actual import queue worker. That worker can keep working at whatever pace it received blocks or justifications to process, since results will be sent on the BufferedLinkSender, which is an unbounded channel.

But you can only advance at your own pace if you rely on a background thread.
I want to give as much flexibility as possible to the implementation, and without `poll_actions` customizing the import queue pretty much requires spawning a background thread.

> I simply dislike what seems like unnecessary complications around the proposed poll_actions method.

This necessity for `BufferedLinkSender`/`BufferedLinkReceiver` is caused by the `JustificationImport::on_start` and `BlockImport::on_start` methods, which are also very poor design in my opinion (initialization is what constructors are for).

Since you have to pass a `&mut dyn Link` to their `on_start` method, I have no option but you give "full link capabilities" to the channel between the worker and the background thread.

If it wasn't for them, you'd just have a channel between the worker and the main thread with just 3 possible messages.

**Comment by gterzian** [comment]
Thanks for the explanation. 

So let's try to be very clear.  We have two traits, and it's how they are defined that will determine whether we gain in flexibility or not:

1. `Link`, that is defined as being `Send`. 
2. `ImportQueue`, which is `Send`(and previously `Sync` as well).

`Link` is how the inport-queue internally communicates back to network, and it's also used by a few other things that the import-queue has access to(`JustificationImport` and `BlockImport`). 

`ImportQueue`, is the interface to the import-queue, used by network. 

Then we have `BufferLinkReceiver` and `BufferedLinkSender`, which are concrete communication mechanism used to communicate with network, and `BufferedLinkSender` implements `Link`, hence that is what the import-queue uses internally to communicate back to network. 

So my first suggestion, "Move  `BufferLink*` to network", relates to not having the import-queue being aware of what it is that implements `Link`. In my opinion you gain flexibility, since `Link` doesn't imply any kind of concurrency(other than requiring it to be `Send`). 

In the same way, network shouldn't be aware what it is that implements `ImportQueue`, it should just calls methods like `import_blocks`, and be done with it. 

So what I strongly dislike about the proposed addition of  a `poll_actions` method to `ImportQueue`, becomes apparent in my opinion to anyone looking at the documentation of that method:

```
/// Polls for actions to perform on the network.
///
/// This method should behave in a way similar to `Future::poll`. It can register the current
/// task and notify later when more actions are ready to be polled. To continue the comparison,
/// it is as if this method always returned `Ok(Async::NotReady)`.
```

What is happening here is that `ImportQueue` suddenly has something to do with "polling for actions to perform on the network", and it also seems very much tied into a "task" type of concurrency context. 

Two problems here:

1. `ImportQueue` should only be about "dumb" communication between network and the import-queue, and not have anything to do with how network runs task.
2. With a method like `poll_actions`, I think you're actually tying the whole things into tasks, futures, and so on. So since the goal is increasing the "concurrency agnosticism" of the whole thing, I think we're going backwards. 

My proposed simple solution to these problems:

1. Move `BufferLink*` to network, ensuring `consensus/common/` knows nothing of what it is that implements `Link`. 
2. Replace the `poll_actions` mechanism, by simply polling the `BufferLinkReceiver` in the network task(like is done with `self.protocol_rx.poll()`, and `self.network_port.poll()`).
3. If we don't like `start`(I don't like it either, since the import-queue has already started earlier), maybe use `set_link(Box<dyn Link<B>>)` as a new method on `ImportQueue`, so that there is a mechanism to pass a `Link` to `ImportQueue`. That way, you can always pass whatever you want to `set_link`, like now a futures sender, and you just poll the receiver in the network task. If you want to pass something containing shared-sate, so be it. What would be the use of something like `poll_actions` if `Link` is implemented by something using locks?
4. Yes it would be even better not to require a `set_link` method, so that a `Link` could simply be passed upon instantiation of `BasicQueue`, however that relates to system-wide setup, and perhaps is best done in a follow-up. 

I propose these changes to the current PR, since I am of the opinion those are better ways to achieve the goal of the PR: more flexibility in terms of concurrency between network and the import-queue.

**Comment by tomaka** [comment]
> So my first suggestion, "Move BufferLink* to network", relates to not having the import-queue being aware of what it is that implements Link. In my opinion you gain flexibility, since Link doesn't imply any kind of concurrency(other than requiring it to be Send).

Again, the fact that the `BasicImportQueue uses `BufferedLink*` is purely an implementation detail. The `ImportQueue` and `BasicImportQueue` are not aware of what is being passed to `poll_actions`.

> Then we have BufferLinkReceiver and BufferedLinkSender, which are concrete communication mechanism used to communicate with network, and BufferedLinkSender implements Link, hence that is what the import-queue uses internally to communicate back to network.

That's true true. The `BufferedLink*` are used only for communication between the import queue worker and the import queue frontend. The network communicates with the import queue frontend and is not aware of the fact that `BufferedLink*` is being used.

> In the same way, network shouldn't be aware what it is that implements ImportQueue, it should just calls methods like import_blocks, and be done with it.

That's true as well, both before and after this PR.

> What is happening here is that ImportQueue suddenly has something to do with "polling for actions to perform on the network"

> ImportQueue should only be about "dumb" communication between network and the import-queue, and not have anything to do with how network runs task.

It's more "polling for actions to perform as an outcome of the import queue importing a block".
I agree that tying it to network actions is not great, but changing that would require modifying `Link`, which is a bit out of scope.

> With a method like poll_actions, I think you're actually tying the whole things into tasks, futures, and so on. So since the goal is increasing the "concurrency agnosticism" of the whole thing, I think we're going backwards.

I'm doing this as a consequence of the fact that `Futures` are getting stable, and that in general asynchronous Rust is become a primary part of idiomatic Rust. In the near future there's not going to be a different anymore between using futures and using iterators for example.

**Comment by gterzian** [comment]
> The BufferedLink* are used only for communication between the import queue worker and the import queue frontend. The network communicates with the import queue frontend and is not aware of the fact that BufferedLink* is being used.

Ok, and the "import-queue front-end" is that which implements `ImportQueue`, which I though of as an interface used by network. Your current approach makes it both something networks uses to communicate with the import-queue, and something the import-queue uses to communicate with network, since "The BufferedLink* are used only for communication between the import queue worker and the import queue frontend", with "the import queue frontend" effectively running in the network task. 

To me BufferedLink* is a mechanism belonging in network, with the sender half implementing `Link` and used by the import-queue to communicate with network, and the receiver half used by network to receive messages. 

I don't understand why you want to entangle `ImportQueue` with a concept that is about communicating back to network, or at least relates to how network receives communication. That seems to me like a sure way to merge two unrelated concepts, and bring unintended consequences(and likely not flexibility).

We can agree to disagree.  

> I'm doing this as a consequence of the fact that Futures are getting stable, and that in general asynchronous Rust is become a primary part of idiomatic Rust.

Fine, at least we're clear about it.

**Comment by tomaka** [comment]
> Ok, and the "import-queue front-end" is that which implements ImportQueue, which I though of as an interface used by network. Your current approach makes it both something networks uses to communicate with the import-queue, and something the import-queue uses to communicate with network, since "The BufferedLink* are used only for communication between the import queue worker and the import queue frontend", with "the import queue frontend" effectively running in the network task.

I think I get it. When I say "the import queue", I'm referring to the `ImportQueue` trait.
The actual worker is private and has always been private, and the fact that we have a worker at all should be an implementation detail of whatever implements the `ImportQueue` trait.

All I want is a clear interface between the import queue and whatever uses it (which happens to be the network at the moment, but could also be the service, and could be unit tests). I don't want someone who uses the `ImportQueue` trait to have to know that there's a background worker.
From the point of view of the user of `ImportQueue`, the "actual import queue" (aka. the worker) doesn't exist/is irrelevant.

> I don't understand why you want to entangle ImportQueue with a concept that is about communicating back to network, or at least relates to how network receives communication. That seems to me like a sure way to merge two unrelated concepts, and bring unintended consequences(and likely not flexibility).

The fact that the import queue is entangled with the network is because of the design of the `Link` trait, and I didn't want to touch this trait as part of this PR.
They are entangled before this PR, after this PR, or with your proposal. It's an orthogonal problem.

File: core/consensus/common/src/import_queue.rs [hunk 2]
```diff
@@ -98,179 +97,83 @@ pub trait Verifier<B: BlockT>: Send + Sync {
...
-/// ensuring blocks in this implementation are imported sequentially and in
-/// order (as received by the "BlockImporter").
-///
-/// As long as the "BasicQueue" is not dropped, the "BlockImporter" will keep
-/// running. The "BlockImporter" owns a sender to the "BlockImportWorker",
-/// ensuring that the worker is kept alive until that sender is dropped.
 impl<B: BlockT> BasicQueue<B> {
 	/// Instantiate a new basic queue, with given verifier.
+	///
+	/// This crates a background thread, and calls `on_start` on the justification importer and
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]:
```suggestion
	/// This creates a background thread, and calls `on_start` on the justification importer and
```

**Comment by gterzian** [comment]
Ok so let's merge this(there are plenty of approvals anyway). 

Just trying to wrap up the discussion because I think I'm starting to get what the misunderstanding is:

- @tomaka it seems to me that you feel that the `start` method means users of `ImportQueue` have to be aware that there's a background thread involved. I don't really agree, since we could rename this to `set_link` or whatever, and `ImportQueue` could be in the same thread and simply be waiting on a link. I think the real problem is that we create the import-queue in one place, and then the link in another, and then there is a need to pass the link to the import-queue one way or another(`poll_actions` still requires passing a link around, for example, so it's pretty similar to `start`).
- I feel that `poll_actions` is unnecessary, since the network task could just poll the buffered receiver. Now off-course that requires passing the sender to the queue at some point, which requires a `start/set_link` method. 
- The real cost of `poll_actions`, and the fact that the "polling" of messages from the queue is becoming part of the `ImportQueue` trait, will become apparent when we need to break-up the network task into smaller tasks. Why? Because the task calling `import_blocks` will probably not be the same that calls `poll_actions`, so you're going to have to share the import-queue across tasks. By polling the receiver instead(and passing the sender along to the queue via another method, or at initialization somehow), one task could just do `buffered_receiver.poll`, and the other would do `import_queue.import_blocks`, without the need to share the queue. 

So in summary, I think the `start` method is perhaps not optimal, but it's less complicated then `poll_actions`, with the latter making the `ImportQueue` trait something not only used to communicate with the queue via methods like `import_blocks`, but also a way to poll for messages from the queue. This might be fine now that everything happens in the same task, my guess is that the unintended consequences will become apparent once you try to break up that task into smaller ones(which is something I have highlighted at other steps of this refactoring). We'll see when we get there, and it is still worth it to consider these things in advance in my opinion.

**Comment by rphmeier** [comment]
> The fact that the import queue is entangled with the network is because of the design of the Link trait, and I didn't want to touch this trait as part of this PR.

At the end of the day, we are going to want to track whether things have gone wrong with import and send it up to the higher level network code so it can note the peer or peers who sent us that data. That means that we'll always send Peer IDs along with the data in some form. That's what `Link` is for, and while not perfectly generic, it's actually necessary in order to write correct import logic.

**Comment by tomaka** [comment]
> At the end of the day, we are going to want to track whether things have gone wrong with import and send it up to the higher level network code so it can note the peer or peers who sent us that data. That means that we'll always send Peer IDs along with the data in some form. That's what Link is for, and while not perfectly generic, it's actually necessary in order to write robust import and networking logic.

What I had in mind by saying that `Link` is entangled with the network, is that to me the import queue should be more similar to "this block import succeeded" or "this block import failed", and shouldn't have methods such as `report_peer`, `request_justification` or `restart`.

**Comment by tomaka** [comment]
I've got at least 3 changes in mind that are blocked on this PR, so would it be ok to merge it?
